File: src/main/java/com/epicsquid/ranchersdelight/RanchersDelight.java
package com.epicsquid.ranchersdelight;

import com.epicsquid.ranchersdelight.config.RanchersDelightConfig;
import com.epicsquid.ranchersdelight.init.RanchersDelightBlockEntities;
import com.epicsquid.ranchersdelight.init.RanchersDelightBlocks;
import com.epicsquid.ranchersdelight.init.RanchersDelightItems;
import com.epicsquid.ranchersdelight.init.RanchersDelightMenus;
import com.tterrag.registrate.Registrate;
import net.minecraft.world.item.CreativeModeTab;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.util.Lazy;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.config.ModConfig;
import org.jetbrains.annotations.NotNull;

@Mod(RanchersDelight.MODID)
public class RanchersDelight {
	public static final String MODID = "ranchersdelight";

	private static final Lazy<Registrate> REGISTRATE = Lazy.of(() -> Registrate.create(MODID));

	public static final CreativeModeTab ITEM_GROUP = new CreativeModeTab(RanchersDelight.MODID) {
		@Override
		@NotNull
		public ItemStack makeIcon() {
			return new ItemStack(RanchersDelightItems.RABBIT_WOOL.get());
		}
	};

	public RanchersDelight() {
		ModLoadingContext.get().registerConfig(ModConfig.Type.SERVER, RanchersDelightConfig.SERVER_CONFIG_SPEC);

		REGISTRATE.get().creativeModeTab(() -> ITEM_GROUP);

		RanchersDelightItems.init();
		RanchersDelightBlocks.init();
		RanchersDelightBlockEntities.init();
		RanchersDelightMenus.init();

		MinecraftForge.EVENT_BUS.register(this);
	}

	public static Registrate registrate() {
		return REGISTRATE.get();
	}
}

File: src/main/kotlin/dev/epicsquid/ranchersdelight/RanchersDelight.kt
package dev.epicsquid.ranchersdelight

import dev.epicsquid.ranchersdelight.config.Config
import dev.epicsquid.ranchersdelight.data.*
import dev.epicsquid.ranchersdelight.registry.BlockEntityRegistry
import dev.epicsquid.ranchersdelight.registry.BlockRegistry
import dev.epicsquid.ranchersdelight.registry.ItemRegistry
import net.minecraft.core.HolderLookup.Provider
import net.minecraftforge.common.data.BlockTagsProvider
import net.minecraftforge.data.event.GatherDataEvent
import net.minecraftforge.fml.ModLoadingContext
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.config.ModConfig.Type.SERVER
import thedarkcolour.kotlinforforge.forge.MOD_BUS

@Mod(RanchersDelight.MODID)
object RanchersDelight {
	const val MODID = "ranchersdelight"

	init {
		val modEventBus = MOD_BUS
		ModLoadingContext.get().registerConfig(SERVER, Config.RANCHERS_DELIGHT_CONFIG_SPEC, "ranchers-delight-server.toml")
		BlockRegistry.REGISTRY.register(modEventBus)
		ItemRegistry.REGISTRY.register(modEventBus)
		BlockEntityRegistry.REGISTRY.register(modEventBus)
	}

	private fun gatherData(event: GatherDataEvent) {
		val generator = event.generator
		val packOutput = event.generator.packOutput
		val lookupProvider = event.lookupProvider
		val existingFileHelper = event.existingFileHelper

		generator.addProvider(event.includeClient(), RanchersDelightBlockStates(packOutput, existingFileHelper))
		generator.addProvider(event.includeClient(), RanchersDelightItemModels(packOutput, existingFileHelper))
		generator.addProvider(event.includeClient(), RanchersDelightLang(packOutput, "en_us"))

		val blockTagsProvider = object : BlockTagsProvider(packOutput, lookupProvider, MODID, existingFileHelper) {
			override fun addTags(pProvider: Provider) {
			}
		}
		generator.addProvider(
			event.includeServer(),
			RanchersDelightItemTags(packOutput, lookupProvider, blockTagsProvider, existingFileHelper)
		)
		generator.addProvider(event.includeServer(), RanchersDelightRecipes(packOutput))

	}
}